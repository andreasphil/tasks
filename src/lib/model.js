export function uid() {
  return Math.round(Math.random() * 10 ** 16)
    .toString()
    .padEnd(16, "0");
}

/**
 * @template {object} T
 * @param {T | undefined} init
 * @returns {T}
 */
export function createModel(init) {
  const now = new Date().getTime();
  init = init ? structuredClone(init) : {};
  return { id: uid(), createdAt: now, updatedAt: now, ...init };
}

/**
 * @param {object} model
 * @param {number} updatedAt
 */
export function touch(model) {
  model.updatedAt = new Date().getTime();
}
