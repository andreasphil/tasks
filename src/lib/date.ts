import dayjs from "dayjs";

/* -------------------------------------------------- *
 * Formatting                                         *
 * -------------------------------------------------- */

export function getDateHint(date: Date | string): string {
  const selected = dayjs(date);
  const day = selected.format("dddd");
  const diff = Math.ceil(selected.diff(dayjs(), "days", true));

  let readableDiff = "today";
  if (diff === -1) readableDiff = "yesterday";
  else if (diff === 1) readableDiff = "tomorrow";
  else if (diff > 1) readableDiff = `in ${diff} days`;
  else if (diff < -1) readableDiff = `${Math.abs(diff)} days ago`;

  return `${day} (${readableDiff})`;
}
