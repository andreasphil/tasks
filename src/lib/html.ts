export function html(raw: TemplateStringsArray, ...values: any[]) {
  return String.raw({ raw }, ...values);
}
