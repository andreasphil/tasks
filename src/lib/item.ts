import {
  parse,
  stringify,
  taskStatuses,
  type DeepWritable,
  type Item,
  type TaskStatus,
  type UncheckedItem,
} from "@/lib/parser";
import dayjs from "dayjs";

/* -------------------------------------------------- *
 * Mutating items                                     *
 * -------------------------------------------------- */

/**
 * Special variant of the Item where all properties that can safely be changed
 * via `asWritable` or `mutate` are writable.
 */
export type WritableItem = Omit<Item, "dueDate" | "status" | "type"> &
  DeepWritable<Pick<Item, "dueDate" | "status" | "type">>;

/**
 * Change the status of the specified item. This will mutate the item.
 *
 * @param item The item to change
 * @param status The next status of the item
 */
function setStatus(item: UncheckedItem, status: TaskStatus): void {
  if (item.type !== "task") return;

  item.status = status;

  const statusToken = item.tokens.find((i) => i.type === "status");
  if (statusToken) {
    const statusChar = taskStatuses[status];
    statusToken.match = statusToken.match.replace(/\[.\]/, `[${statusChar}]`);
    statusToken.text = statusChar;
  }

  item.raw = stringify(item);
}

/**
 * Change the due date of the specified item. This will mutate the item.
 *
 * @param item The item to change
 * @param date The next due date of the item. Will remove an existing due
 *  date then set to undefined.
 */
function setDueDate(item: UncheckedItem, date?: Date): void {
  const hasDueDate = item.dueDate !== undefined;
  let newRaw = item.raw;
  const newDueDateStr = date ? dayjs(date).format("YYYY-MM-DD") : "";

  // 1. No due date before, no due date after @ skip
  if (!hasDueDate && !date) return;
  // 2. No due date before, due date after @ add
  else if (!hasDueDate && date) {
    newRaw = item.raw.replace(/ ?$/, ` @${newDueDateStr}`);
  }
  // 3. Due date before, no due date after @ remove
  else if (hasDueDate && !date) {
    const dueDateToken = item.tokens.find((i) => i.type === "dueDate");
    if (!dueDateToken) return;
    const exp = new RegExp(`\\s?@${dueDateToken.text}`);
    newRaw = newRaw.replace(exp, "");
  }
  // 4. Due date before, due date after @ update
  else if (hasDueDate && date) {
    const dueDateToken = item.tokens.find((i) => i.type === "dueDate");
    if (!dueDateToken) return;
    newRaw = newRaw.replace(`@${dueDateToken.text}`, `@${newDueDateStr}`);
  }

  Object.assign(item, parse(newRaw));
}

/**
 * Change the status of the specified item. This will mutate the item.
 *
 * @param item The item to change
 * @param type The next type of the item
 */
function setType(item: UncheckedItem, type: UncheckedItem["type"]): void {
  if (item.type === type) return;

  let newRaw = item.raw;

  // Normalize the type of the item by removing any type specific content
  if (item.type === "task") newRaw = newRaw.replace(/^(\s*)\[.\] /, "$1");
  else if (item.type === "heading") newRaw = newRaw.replace(/^# /, "");
  else if (item.type === "note") newRaw = newRaw.replace(/^(\s*)\[-\] /, "$1");

  // Add type specific content
  if (type === "task") newRaw = newRaw.replace(/^(\s*)/, "$1[ ] ");
  else if (type === "heading") newRaw = `# ${newRaw.trimStart()}`;

  Object.assign(item, parse(newRaw));
  item.status = type === "task" ? "incomplete" : null;
}

/**
 * Returns a wrtiable proxy to the original item. This will allow changing some
 * of the properties of an item while keeping the data structure intact and
 * consistent, i.e. all tokens, raw value, and dependencies between properties
 * of the item will also be updated to match the new value of the property.
 *
 * @param item The item that should be changed.
 * @returns A writable proxy to the original item. Note that this is not a
 *  copy. Any changes made to the writable item will change the original item.
 */
export function asWritable(item: Item): WritableItem {
  return new Proxy(structuredClone(item as UncheckedItem), {
    set: (target, property, value) => {
      if (property === "type") {
        setType(target, value);
      } else if (property === "dueDate") {
        setDueDate(target, value);
      } else if (property === "status") {
        setStatus(target, value);
      } else {
        const propertyName = String(property);
        throw new Error(`Setting "${propertyName}" is not currently supported`);
      }

      return true;
    },
  });
}

/**
 * Allows you to change an item in place by writing all changes you make in the
 * `factory` callback back into the item.
 *
 * @param item The item that should be mutated.
 * @param factory A function making the changes to the item.
 */
export function mutate(
  item: Item,
  factory: (item: WritableItem) => void
): void {
  const w = asWritable(item);
  factory(w);
  Object.assign(item, w);
}

/* -------------------------------------------------- *
 * Sorting items                                      *
 * -------------------------------------------------- */

const statusWeights: Record<TaskStatus, number> = {
  important: 10000,
  inProgress: 1000,
  incomplete: 100,
  question: 10,
  completed: 1,
} as const;

/**
 * Sorts 2 items by the following criteria:
 *
 * 1. Sections and notes are always equal, i.e. comparing two should not change
 *    their order.
 * 2. When comparing sections or notes with tasks, the order should also stay
 *    the same.
 * 3. Tasks are first sorted by their status (important, in progress, open,
 *    question, done), then due dates (ascending; tasks with due dates should
 *    come before tasks without due dates). Tasks with the same due date or
 *    status should not change their order.
 * 4. To accomodate nested tasks, white space at the beginning of the line
 *    always takes precedence over all other criteria. Tasks with different
 *    leading white space should never change order. Note that sorting an
 *    entire page using `compare(item, item)` will mess up pages with nested
 *    items, as those nested items need to be moved together with their parent
 *    item. If you need to sort an entire page, use `sort(page)`.
 * 5. The order of items and blank lines should not change.
 *
 * Note that this function is only suitable for comparing two individual items.
 * It should not be used for sorting an entire page as it will mess up the
 * nesting of tasks. To sort a page, use `page.ts@sort()` instead.
 *
 * @param a
 * @param b
 */
export function compare(a: Item, b: Item): number {
  // Keep order of items of different types
  if (a.type !== b.type || a.type !== "task" || b.type !== "task") return 0;

  // Keep order of items with different leading white space
  const aIndent = a.raw.match(/^\s*/)?.[0] ?? "";
  const bIndent = b.raw.match(/^\s*/)?.[0] ?? "";
  if (aIndent !== bIndent) return 0;

  // Sort by status
  if (a.status !== b.status) {
    return statusWeights[b.status] - statusWeights[a.status];
  }

  // Sort by due date
  if (a.dueDate && !b.dueDate) return -1;
  else if (b.dueDate && !a.dueDate) return 1;
  else {
    const aDueDate = dayjs((a as UncheckedItem).dueDate).startOf("day");
    const bDueDate = dayjs((b as UncheckedItem).dueDate).startOf("day");
    const diff = aDueDate.diff(bDueDate);
    if (diff !== 0) return diff;
  }

  // Default: keep the order
  return 0;
}
