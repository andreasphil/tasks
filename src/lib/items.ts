import {
  Item,
  TaskStatus,
  UncheckedItem,
  stringify,
  taskStatuses,
} from "@/lib/parser";
import { DeepWritable } from "./utils";

/**
 * Special variant of the Item where all properties that can safely be changed
 * via `asWritable` or `mutate` are writable.
 */
export type WritableItem = Omit<Item, "dueDate" | "status"> &
  DeepWritable<Pick<Item, "dueDate" | "status">>;

function setStatus(item: UncheckedItem, newStatus: TaskStatus) {
  if (item.type !== "task") return;

  item.status = newStatus;

  const statusToken = item.tokens.find((i) => i.type === "status");
  if (statusToken) {
    const statusChar = taskStatuses[newStatus];
    statusToken.match = statusToken.match.replace(/\[.\]/, `[${statusChar}]`);
    statusToken.text = statusChar;
  }

  item.raw = stringify(item);
}

/**
 * Returns a wrtiable proxy to the original item. This will allow changing some
 * of the properties of an item while keeping the data structure intact and
 * consistent, i.e. all tokens, raw value, and dependencies between properties
 * of the item will also be updated to match the new value of the property.
 */
export function asWritable(item: Item): WritableItem {
  return new Proxy(structuredClone(item as UncheckedItem), {
    set: (target, property, value) => {
      if (property === "dueDate") {
        // TODO: Implement
        throw new Error("Setting dueDate is not currently supported");
      } else if (property === "status") {
        setStatus(target, value);
      } else {
        const propertyName = String(property);
        throw new Error(`Setting "${propertyName}" is not currently supported`);
      }

      return true;
    },
  });
}

/**
 * Allows you to change an item in place by writing all changes you make in the
 * `factory` callback back into the item.
 */
export function mutate(item: Item, factory: (item: WritableItem) => void) {
  const w = asWritable(item);
  factory(w);
  Object.assign(item, w);
}
