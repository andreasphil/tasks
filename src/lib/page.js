/**
 * @typedef {object} Page
 * @prop {string} id
 * @prop {number} createdAt
 * @prop {number} updatedAt
 * @prop {string} text
 */

/**
 * @param {Object} page
 * @param {string} page.text
 */
export function getTitle({ text }) {
  const [firstLine] = text.trim().split("\n");
  return firstLine || "Untitled";
}
