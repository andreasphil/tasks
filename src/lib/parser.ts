/**
 * This is the parser for the converting a string into a task. The format for
 * tasks is loosely based on the Bullet Journal notation, and is defined as
 * follows:
 *
 * Lines that start with a single character enclosed in square brackets are
 * tasks. The character inside the brackets defines the status of the task.
 * The following statuses are supported:
 *
 * [ ] Incomplete task
 * [x] Completed
 * [/] In progress
 * [*] Important
 * [-] Note
 * [?] Question
 *
 * The text after the brackets is the text of the task. Lines can also start
 * with whitespace, which is ignored (but can be used for indentation). The
 * text of the task can contain tags and a due date:
 *
 * - Tags: #tag1 #tag2 (tags can be any letter, number, or underscore)
 * - Due date: ->2021-01-01
 *
 * Besides tasks, lists can also contain notes and sections. Sections are started
 * by headings, which are lines that start with a single #.
 *
 * Everything else is considered a note. Notes have no special syntax or meaning,
 * but are used to add additional information to a task. They are also allowed
 * to contain tags and due dates.
 */

import { DeepReadonly } from "./utils";

/* -------------------------------------------------- *
 * Types                                              *
 * -------------------------------------------------- */

export const taskStatuses = {
  incomplete: " ",
  completed: "x",
  inProgress: "/",
  important: "*",
  question: "?",
} as const;

const statusChars = Object.values(taskStatuses);

/** Represents the status of a task. */
export type TaskStatus = keyof typeof taskStatuses;

/** Different types of tokens that can be found in an item. */
export type TokenType =
  | "headingMarker"
  | "status"
  | "dueDate"
  | "tag"
  | "link"
  | "text";

/** Represents a token in an item. */
export type Token = {
  type: TokenType;
  text: string;
  match: string;
  matchStart: number;
  matchLength: number;
};

/**
 * Represents a parsed item. "Item" is the generic type for anything that can
 * be included in a list, such as a task, note, or section heading.
 *
 * Unchecked items allow changing any property in any way you want. Note that
 * since properties depend on each other (e.g. status is parsed from raw, and
 * also leads to the presence of a "status" token), changing them individually
 * might result in inconsistent data within the item.
 */
export type UncheckedItem = {
  raw: string;
  text: string;
  tokens: Token[];
  tags: string[];
  dueDate?: Date;
} & (
  | { type: "note" | "heading"; status: null }
  | { type: "task"; status: TaskStatus }
);

/**
 * Represents a parsed item. "Item" is the generic type for anything that can
 * be included in a list, such as a task, note, or section heading.
 *
 * Note that in a regular item, all properties are readonly. This is because
 * some depend on each other (e.g. status is parsed from raw, and also leads
 * to the presence of a "status" token), and changing them individually might
 * result in inconsistent data within the item.
 *
 * Check out `items.ts@asWritable` to learn how to safely mutate items.
 */
export type Item = DeepReadonly<UncheckedItem>;

/* -------------------------------------------------- *
 * Regexes                                            *
 * -------------------------------------------------- */

// Singular means only one of these can be matched per line. Plural means
// multiple matches can be found per line.
const regexes = {
  // Section headings
  heading: /^# /,
  // Task status
  status: new RegExp(
    `(?<=^[^\\S\\n]*)\\[(?<status>[${statusChars.join("")}])\\]`
  ),
  // Due date
  dueDate: /->(?<dueDate>\d{4}-\d{2}-\d{2})/,
  // Tags
  tags: /#(?<tag>\w+)/g,
  // Links
  links: /(?<url>https?:\/\/\S+)/g,
} as const;

const mergedRegex = new RegExp(
  Object.values(regexes)
    .map((i) => `(${i.source})`)
    .join("|"),
  "gm"
);

/* -------------------------------------------------- *
 * Utilities                                          *
 * -------------------------------------------------- */

/**
 * Converts a status character to a TaskStatus. If the character is not a valid
 * status, 'incomplete' is returned as a fallback. If the `strict` parameter is
 * set to true, an error is thrown instead.
 */
function toStatus(status: string, strict = false): TaskStatus {
  let result: TaskStatus = "incomplete";

  switch (status) {
    case taskStatuses.completed:
      result = "completed";
      break;
    case taskStatuses.inProgress:
      result = "inProgress";
      break;
    case taskStatuses.important:
      result = "important";
      break;
    case taskStatuses.question:
      result = "question";
      break;
    default:
      if (strict) throw new Error(`Invalid status: ${status}`);
      break;
  }

  return result;
}

/**
 * Returns a copy of the given array of tokens, sorted by their occurrence in
 * the original string (i.e. the `matchStart` property).
 */
function sortTokens(tokens: Token[]): Token[] {
  return [...tokens].sort((a, b) => a.matchStart - b.matchStart);
}

/**
 * For a given string and an array of tokens that occur in that string, this
 * function fills the gaps between those previously found tokens with the text
 * that was skipped. This is necessary because the regexes used to find the
 * tokens don't match the entire string, but only the parts that are relevant
 * to the token type. To reconstruct the original string, we need to fill in
 * the gaps between the tokens.
 */
function insertTextTokens(original: string, tokens: Token[]): Token[] {
  // If there are no tokens, the entire string is a text token, so we return
  // that and exit early.
  if (tokens.length === 0) {
    return [
      {
        type: "text",
        text: original,
        match: original,
        matchStart: 0,
        matchLength: original.length,
      },
    ];
  }

  const sortedTokens = sortTokens(tokens);
  let result: Token[] = [];
  let cursor = 0;
  let buffer = "";
  let nextToken = sortedTokens.shift();

  while (cursor < original.length) {
    // If there is no next token, we push the rest of the string as a text
    // token and we're done.
    if (!nextToken) {
      buffer = original.slice(cursor);
      result.push({
        type: "text",
        text: buffer,
        match: buffer,
        matchStart: cursor,
        matchLength: buffer.length,
      });

      break;
    }

    // If the cursor is at the start of the next token, we first flush the
    // buffer by adding it as a text token, then push the next token to the
    // result and move the cursor to the end of the token. We can then take
    // a new token from the queue.
    else if (cursor === nextToken.matchStart) {
      if (buffer.length > 0) {
        result.push({
          type: "text",
          text: buffer,
          match: buffer,
          matchStart: cursor - buffer.length,
          matchLength: buffer.length,
        });
        buffer = "";
      }

      result.push(nextToken);
      cursor += nextToken.matchLength;
      nextToken = sortedTokens.shift();
    }

    // If the cursor doesn't currently point to a token, we push the current
    // character to the buffer and move the cursor forward.
    else {
      buffer += original[cursor];
      cursor++;
    }
  }

  return result;
}

/* -------------------------------------------------- *
 * Parser                                             *
 * -------------------------------------------------- */

export function parse(input: string): Item {
  let token = null;

  let type: UncheckedItem["type"] = "note";
  let text = input;
  let tags: Set<Item["tags"][number]> = new Set();
  let dueDate: UncheckedItem["dueDate"] = undefined;
  let status: TaskStatus = "incomplete";
  let tokens: Token[] = [];

  while ((token = mergedRegex.exec(input))) {
    const tokenDescriptor: Token = {
      type: "text",
      text: token[0],

      match: token[0],
      matchStart: token.index,
      matchLength: token[0].length,
    };

    if (token[1]) {
      // This is a section heading (but it doesn't have a named group in the
      // regex, so we check for the group index instead).
      type = "heading";
      text = input.replace(regexes.heading, "");
      tokenDescriptor.type = "headingMarker";
      tokenDescriptor.text = token[1];
    } else if (token.groups?.status) {
      // Tasks, so we'll also need to check for the status
      type = "task";
      text = input.replace(regexes.status, "");
      status = toStatus(token.groups.status);
      tokenDescriptor.type = "status";
      tokenDescriptor.text = token.groups.status;
    } else if (token.groups?.dueDate && !dueDate) {
      // Due date
      const maybeDueDate = new Date(token.groups.dueDate);
      if (!Number.isNaN(maybeDueDate.getTime())) {
        dueDate = maybeDueDate;
        tokenDescriptor.type = "dueDate";
        tokenDescriptor.text = token.groups.dueDate;
      }
    } else if (token.groups?.tag) {
      // Tags
      tags.add(token.groups.tag);
      tokenDescriptor.type = "tag";
      tokenDescriptor.text = token.groups.tag;
    } else if (token.groups?.url) {
      // Links
      tokenDescriptor.type = "link";
      tokenDescriptor.text = token.groups.url;
    }

    tokens.push(tokenDescriptor);
  }

  tokens = insertTextTokens(input, tokens);

  // @ts-expect-error Complains about the missing status, but we'll add it later
  // if needed.
  const result: UncheckedItem = {
    type,
    raw: input,
    text,
    tokens,
    tags: Array.from(tags),
    dueDate,
  };

  if (result.type === "task") result.status = status;

  return result as Item;
}

/* -------------------------------------------------- *
 * Stringifier                                        *
 * -------------------------------------------------- */

export function stringify(item: Item): string {
  return item.tokens.map((i) => i.match).join("");
}
