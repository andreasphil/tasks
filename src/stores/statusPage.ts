import { compare, mutate } from "@/lib/item";
import { getTitle } from "@/lib/page";
import { TaskStatus, type Item } from "@/lib/parser";
import { computed } from "vue";
import { usePages } from "./pages";

export type StatusPageItem = Item & {
  pageName: string;
};

export function useStatusPage() {
  const { pages, updatePage } = usePages();

  /* -------------------------------------------------- *
   * Items                                              *
   * -------------------------------------------------- */

  type UpdateFn = (factory: Parameters<typeof mutate>[1]) => void;
  type UpdateableItem = StatusPageItem & { update?: UpdateFn };

  function updateFn(page: string, index: number): UpdateFn {
    return (factory) => {
      if (!pages[page]) return;
      const newItems = [...pages[page].items];
      newItems[index] = mutate(newItems[index], factory);
      updatePage(page, { items: newItems });
    };
  }

  function updateItem(index: number, from: TaskStatus, to: TaskStatus): void {
    const updateFn = items.value.get(from)?.at(index)?.update;
    updateFn?.((item) => {
      item.status = to;
    });
  }

  /**
   * The list of items on this page, where each item also has a function to
   * update the item. The function keeps track of the page and index of the
   * item. The function is optional and doesn't exist for items that can't
   * be updated.
   */
  const items = computed(() => {
    const buffer = new Map<TaskStatus, UpdateableItem[]>();
    buffer.set("completed", []);
    buffer.set("important", []);
    buffer.set("incomplete", []);
    buffer.set("inProgress", []);
    buffer.set("question", []);

    Object.values(pages)
      // Get items with update function from all pages
      .flatMap((page) =>
        page.items.map((item, i) => ({
          ...item,
          raw: item.raw.replace(/^\t+/, ""),
          pageName: getTitle(page),
          update: updateFn(page.id, i),
        }))
      )

      // Remove items other than tasks
      .filter((item) => item.type === "task" && item.status !== null)

      // Group by status
      .reduce((all, item) => {
        all.get(item.status!)!.push(item);
        all.get(item.status!)!.sort(compare);
        return all;
      }, buffer);

    return buffer;
  });

  /* -------------------------------------------------- *
   * Page text                                          *
   * -------------------------------------------------- */

  return { items, updateOnPage: updateItem };
}
