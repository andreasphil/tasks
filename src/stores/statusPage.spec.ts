import { parse } from "@/lib/parser";
import { afterEach, describe, expect, test, vi } from "vitest";
import { useStatusPage } from "./statusPage";

const mocks = vi.hoisted(() => {
  return { updatePage: vi.fn() };
});

vi.mock("@/stores/pages", () => ({
  usePages: () => ({
    pages: {
      foo: {
        id: "foo",
        items: [
          "Page 1",
          "[ ] Open from Page 1",
          "[/] In progress from Page 1",
          "[?] Blocked from Page 1",
          "[*] Important from Page 1",
          "[x] Completed from Page 1",
        ].map(parse),
      },

      bar: {
        id: "bar",
        items: [
          "Page 2",
          "[ ] Open from Page 2",
          "[/] In progress from Page 2",
          "[?] Blocked from Page 2",
          "[*] Important from Page 2",
          "[x] Completed from Page 2",
        ].map(parse),
      },
    },
    updatePage: mocks.updatePage,
  }),
}));

describe("statusPage", () => {
  afterEach(() => {
    vi.resetAllMocks();
  });

  test("returns tasks grouped by status", () => {
    const { items } = useStatusPage();
    expect(items.value).toMatchSnapshot();
  });

  test("updates an item on the page", () => {
    const { updateOnPage } = useStatusPage();

    updateOnPage(0, "incomplete", "inProgress");

    expect(mocks.updatePage).toHaveBeenNthCalledWith(
      1,
      "foo",
      expect.objectContaining({
        items: expect.arrayContaining([
          expect.objectContaining({ raw: "[/] Open from Page 1" }),
        ]),
      })
    );

    updateOnPage(1, "important", "completed");

    expect(mocks.updatePage).toHaveBeenNthCalledWith(
      2,
      "bar",
      expect.objectContaining({
        items: expect.arrayContaining([
          expect.objectContaining({ raw: "[x] Important from Page 2" }),
        ]),
      })
    );
  });
});
