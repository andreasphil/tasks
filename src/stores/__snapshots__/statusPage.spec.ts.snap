// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`statusPage > returns tasks grouped by status 1`] = `
Map {
  "completed" => [
    {
      "dueDate": undefined,
      "pageName": "Page 1",
      "raw": "[x] Completed from Page 1",
      "status": "completed",
      "tags": [],
      "text": " Completed from Page 1",
      "tokens": [
        {
          "match": "[x]",
          "matchLength": 3,
          "matchStart": 0,
          "text": "x",
          "type": "status",
        },
        {
          "match": " Completed from Page 1",
          "matchLength": 22,
          "matchStart": 3,
          "text": " Completed from Page 1",
          "type": "text",
        },
      ],
      "type": "task",
      "update": [Function],
    },
    {
      "dueDate": undefined,
      "pageName": "Page 2",
      "raw": "[x] Completed from Page 2",
      "status": "completed",
      "tags": [],
      "text": " Completed from Page 2",
      "tokens": [
        {
          "match": "[x]",
          "matchLength": 3,
          "matchStart": 0,
          "text": "x",
          "type": "status",
        },
        {
          "match": " Completed from Page 2",
          "matchLength": 22,
          "matchStart": 3,
          "text": " Completed from Page 2",
          "type": "text",
        },
      ],
      "type": "task",
      "update": [Function],
    },
  ],
  "important" => [
    {
      "dueDate": undefined,
      "pageName": "Page 1",
      "raw": "[*] Important from Page 1",
      "status": "important",
      "tags": [],
      "text": " Important from Page 1",
      "tokens": [
        {
          "match": "[*]",
          "matchLength": 3,
          "matchStart": 0,
          "text": "*",
          "type": "status",
        },
        {
          "match": " Important from Page 1",
          "matchLength": 22,
          "matchStart": 3,
          "text": " Important from Page 1",
          "type": "text",
        },
      ],
      "type": "task",
      "update": [Function],
    },
    {
      "dueDate": undefined,
      "pageName": "Page 2",
      "raw": "[*] Important from Page 2",
      "status": "important",
      "tags": [],
      "text": " Important from Page 2",
      "tokens": [
        {
          "match": "[*]",
          "matchLength": 3,
          "matchStart": 0,
          "text": "*",
          "type": "status",
        },
        {
          "match": " Important from Page 2",
          "matchLength": 22,
          "matchStart": 3,
          "text": " Important from Page 2",
          "type": "text",
        },
      ],
      "type": "task",
      "update": [Function],
    },
  ],
  "incomplete" => [
    {
      "dueDate": undefined,
      "pageName": "Page 1",
      "raw": "[ ] Open from Page 1",
      "status": "incomplete",
      "tags": [],
      "text": " Open from Page 1",
      "tokens": [
        {
          "match": "[ ]",
          "matchLength": 3,
          "matchStart": 0,
          "text": " ",
          "type": "status",
        },
        {
          "match": " Open from Page 1",
          "matchLength": 17,
          "matchStart": 3,
          "text": " Open from Page 1",
          "type": "text",
        },
      ],
      "type": "task",
      "update": [Function],
    },
    {
      "dueDate": undefined,
      "pageName": "Page 2",
      "raw": "[ ] Open from Page 2",
      "status": "incomplete",
      "tags": [],
      "text": " Open from Page 2",
      "tokens": [
        {
          "match": "[ ]",
          "matchLength": 3,
          "matchStart": 0,
          "text": " ",
          "type": "status",
        },
        {
          "match": " Open from Page 2",
          "matchLength": 17,
          "matchStart": 3,
          "text": " Open from Page 2",
          "type": "text",
        },
      ],
      "type": "task",
      "update": [Function],
    },
  ],
  "inProgress" => [
    {
      "dueDate": undefined,
      "pageName": "Page 1",
      "raw": "[/] In progress from Page 1",
      "status": "inProgress",
      "tags": [],
      "text": " In progress from Page 1",
      "tokens": [
        {
          "match": "[/]",
          "matchLength": 3,
          "matchStart": 0,
          "text": "/",
          "type": "status",
        },
        {
          "match": " In progress from Page 1",
          "matchLength": 24,
          "matchStart": 3,
          "text": " In progress from Page 1",
          "type": "text",
        },
      ],
      "type": "task",
      "update": [Function],
    },
    {
      "dueDate": undefined,
      "pageName": "Page 2",
      "raw": "[/] In progress from Page 2",
      "status": "inProgress",
      "tags": [],
      "text": " In progress from Page 2",
      "tokens": [
        {
          "match": "[/]",
          "matchLength": 3,
          "matchStart": 0,
          "text": "/",
          "type": "status",
        },
        {
          "match": " In progress from Page 2",
          "matchLength": 24,
          "matchStart": 3,
          "text": " In progress from Page 2",
          "type": "text",
        },
      ],
      "type": "task",
      "update": [Function],
    },
  ],
  "question" => [
    {
      "dueDate": undefined,
      "pageName": "Page 1",
      "raw": "[?] Blocked from Page 1",
      "status": "question",
      "tags": [],
      "text": " Blocked from Page 1",
      "tokens": [
        {
          "match": "[?]",
          "matchLength": 3,
          "matchStart": 0,
          "text": "?",
          "type": "status",
        },
        {
          "match": " Blocked from Page 1",
          "matchLength": 20,
          "matchStart": 3,
          "text": " Blocked from Page 1",
          "type": "text",
        },
      ],
      "type": "task",
      "update": [Function],
    },
    {
      "dueDate": undefined,
      "pageName": "Page 2",
      "raw": "[?] Blocked from Page 2",
      "status": "question",
      "tags": [],
      "text": " Blocked from Page 2",
      "tokens": [
        {
          "match": "[?]",
          "matchLength": 3,
          "matchStart": 0,
          "text": "?",
          "type": "status",
        },
        {
          "match": " Blocked from Page 2",
          "matchLength": 20,
          "matchStart": 3,
          "text": " Blocked from Page 2",
          "type": "text",
        },
      ],
      "type": "task",
      "update": [Function],
    },
  ],
}
`;
